File: cpp/oak/vio_replay.cpp
#include <spectacularAI/vio.hpp>
#include <spectacularAI/replay.hpp>

#include <iostream>
#ifdef EXAMPLE_USE_OPENCV
#include <opencv2/opencv.hpp>
#endif

int main(int argc, char** argv) {
    if (argc < 2) return 1;

    std::string dataFolder = argv[1];

    spectacularAI::Vio::Builder vioBuilder = spectacularAI::Vio::builder();
    auto replayApi = spectacularAI::Replay::builder(dataFolder, vioBuilder)
        .build();

    replayApi->setExtendedOutputCallback([&](spectacularAI::VioOutputPtr output, spectacularAI::FrameSet frames) {
        std::cout << output->asJson() << std::endl;
        #ifdef EXAMPLE_USE_OPENCV
        for (int i = 0; i < frames.size(); i++) {
            // Note: typically not main thread
            auto &frame = frames[i];
            if (frame->image) {
                cv::Mat img = frame->image->asOpenCV();
                cv::imshow("Video " + std::to_string(i), img);
                cv::waitKey(1);
            }
        }
        #else
        (void)frames;
        #endif
    });

    replayApi->startReplay();

    return 0;
}

File: python/oak/vio_replay.py
import argparse
import spectacularAI
import cv2

if __name__ == '__main__':
    p = argparse.ArgumentParser(__doc__)
    p.add_argument("dataFolder", help="Folder containing the recorded session for replay", default="data")
    p.add_argument("--preview", help="Show latest primary image as a preview", action="store_true")
    args =  p.parse_args()

    def onOutput(output, frameSet):
        for frame in frameSet:
            if args.preview and frame.image:
                cv2.imshow("Camera #{}".format(frame.index), cv2.cvtColor(frame.image.toArray(), cv2.COLOR_RGB2BGR))
                cv2.waitKey(1)
        print(output.asJson())

    replay = spectacularAI.Replay(args.dataFolder)
    # If frameSet isn't used, it's better to use setOutputCallback(...) which is lighter
    # replay.setOutputCallback(onOutput)
    replay.setExtendedOutputCallback(onOutput)
    replay.runReplay()
